diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
index fea9e3f99..94300221a 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
@@ -1120,6 +1120,7 @@
 			reg = <0x01ca0000 0x1000>;
 			interrupts = <GIC_SPI 89 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&ccu CLK_BUS_MIPI_DSI>;
+			clock-names = "bus";
 			resets = <&ccu RST_BUS_MIPI_DSI>;
 			phys = <&dphy>;
 			phy-names = "dphy";
