diff --git a/scripts/package/mkdebian b/scripts/package/mkdebian
index 8f9b0414a..fc072c707 100755
--- a/scripts/package/mkdebian
+++ b/scripts/package/mkdebian
@@ -202,7 +202,6 @@ Architecture: $debarch
 Description: Linux kernel debugging symbols for $version
  This package will come in handy if you need to debug the kernel. It provides
  all the necessary debug symbols for the kernel and its modules.
-EOF
 
 Package: $dtb_packagename
 Architecture: $debarch
