diff --git a/drivers/gpu/drm/sun4i/sun4i_dotclock.c b/drivers/gpu/drm/sun4i/sun4i_dotclock.c
index 3d0638a3d..4e2f6f49e 100644
--- a/drivers/gpu/drm/sun4i/sun4i_dotclock.c
+++ b/drivers/gpu/drm/sun4i/sun4i_dotclock.c
@@ -99,7 +99,7 @@ static long sun4i_dclk_round_rate(struct clk_hw *hw, unsigned long rate,
 		rounded = clk_hw_round_rate(clk_hw_get_parent(hw),
 					    ideal);
 
-		printk("ideal = %lld, rounded = %d\n", ideal, rounded);
+		printk("ideal = %lld, rounded = %ld\n", ideal, rounded);
 		if (rounded == ideal) {
 			best_parent = rounded;
 			best_div = i;
