diff --git a/drivers/gpu/drm/sun4i/sun6i_mipi_dsi.c b/drivers/gpu/drm/sun4i/sun6i_mipi_dsi.c
index 76bda71ac..a9268ad8a 100644
--- a/drivers/gpu/drm/sun4i/sun6i_mipi_dsi.c
+++ b/drivers/gpu/drm/sun4i/sun6i_mipi_dsi.c
@@ -396,7 +396,6 @@ static u16 sun6i_dsi_get_drq_edge0(struct sun6i_dsi *dsi,
 	dclk_parent_rate = clk_get_rate(clk_get_parent(tcon->dclk));
 	tcon0_div = dclk_parent_rate / dclk_rate;
 
-+	printk("sun6i_dsi_get_drq_edge0: tcon0_div = %ld\n", tcon0_div);
 	edge0 += (mode->hdisplay + 40) * tcon0_div / 8;
 
 	if (edge0 > line_num)
@@ -436,8 +435,6 @@ static void sun6i_dsi_setup_burst(struct sun6i_dsi *dsi,
 		edge1 = sun6i_dsi_get_drq_edge1(dsi, mode, line_num);
 		edge0 = sun6i_dsi_get_drq_edge0(dsi, mode, line_num, edge1);
 
-		printk("edge0 = %d, edge1 = %d, line_num = %d\n",
-			edge0, edge1, line_num);
 		regmap_write(dsi->regs, SUN6I_DSI_BURST_DRQ_REG,
 			     SUN6I_DSI_BURST_DRQ_EDGE0(edge0) |
 			     SUN6I_DSI_BURST_DRQ_EDGE1(edge1));
@@ -474,7 +471,6 @@ static void sun6i_dsi_setup_inst_loop(struct sun6i_dsi *dsi,
 		delay -= 50;
 	}
 
-	printk("sun6i_dsi_setup_inst_loop: delay = %d\n", delay);
 	regmap_write(dsi->regs, SUN6I_DSI_INST_LOOP_SEL_REG,
 		     2 << (4 * DSI_INST_ID_LP11) |
 		     3 << (4 * DSI_INST_ID_DLY));
@@ -607,9 +603,6 @@ static void sun6i_dsi_setup_timings(struct sun6i_dsi *dsi,
 		vblk = 0;
 	}
 
-	printk("hsa = %d, hbp = %d, hfp = %d, hblk = %d, vblk = %d\n",
-		hsa, hbp, hfp, hblk, vblk);
-
 	/* How many bytes do we need to send all payloads? */
 	bytes = max_t(size_t, max(max(hfp, hblk), max(hsa, hbp)), vblk);
 	buffer = kmalloc(bytes, GFP_KERNEL);
@@ -741,7 +734,6 @@ static void sun6i_dsi_encoder_enable(struct drm_encoder *encoder)
 	pm_runtime_get_sync(dsi->dev);
 
 	delay = sun6i_dsi_get_video_start_delay(dsi, mode);
-	printk("sun6i_dsi_get_video_start_delay: delay = %d\n", delay);
 	regmap_write(dsi->regs, SUN6I_DSI_BASIC_CTL1_REG,
 		     SUN6I_DSI_BASIC_CTL1_VIDEO_ST_DELAY(delay) |
 		     SUN6I_DSI_BASIC_CTL1_VIDEO_FILL |
