diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
index bf38fd6d7..b09ff9eae 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
@@ -1,5 +1,4 @@
-/*
- * Copyright (c) 2017 Icenowy Zheng <icenowy@aosc.xyz>
+ /* Copyright (c) 2017 Icenowy Zheng <icenowy@aosc.xyz>
  *
  * Based on sun50i-a64-pine64.dts, which is:
  *   Copyright (c) 2016 ARM Ltd.
@@ -163,7 +162,7 @@
 &i2c0 {
 	status = "okay";
 	
-	gt911@5d {
+/*	gt911@5d {
 		compatible = "goodix,gt911";
 		reg = <0x5d>;
 		
@@ -172,10 +171,10 @@
 		interrupt-parent = <&pio>;
 		interrupts = <7 4 IRQ_TYPE_EDGE_RISING>;	
 		irq-gpios = <&pio 7 4>;/* GPIO_ACTIVE_HIGH>;*/
-		reset-gpios = <&pio 7 11>;/* GPIO_ACTIVE_HIGH>;*/
+/*		reset-gpios = <&pio 7 11>;/* GPIO_ACTIVE_HIGH>;*/
 		/*touchscreen-inverted-x;*/
 		/*touchscreen-inverted-y;*/
-	};	
+/*	};	*/
 };
 
 
