diff --git a/include/trace/events/smbus.h b/include/trace/events/smbus.h
index a4892a187..d2fb6e1d3 100644
--- a/include/trace/events/smbus.h
+++ b/include/trace/events/smbus.h
@@ -138,9 +138,9 @@ TRACE_EVENT_CONDITION(smbus_reply,
 	TP_PROTO(const struct i2c_adapter *adap,
 		 u16 addr, unsigned short flags,
 		 char read_write, u8 command, int protocol,
-		 const union i2c_smbus_data *data, int res),
-	TP_ARGS(adap, addr, flags, read_write, command, protocol, data, res),
-	TP_CONDITION(res >= 0 && read_write == I2C_SMBUS_READ),
+		 const union i2c_smbus_data *data),
+	TP_ARGS(adap, addr, flags, read_write, command, protocol, data),
+	TP_CONDITION(read_write == I2C_SMBUS_READ),
 	TP_STRUCT__entry(
 		__field(int,	adapter_nr		)
 		__field(__u16,	addr			)
