diff --git a/drivers/i2c/i2c-core-slave.c b/drivers/i2c/i2c-core-slave.c
index 47a9f70a2..4a78c65e9 100644
--- a/drivers/i2c/i2c-core-slave.c
+++ b/drivers/i2c/i2c-core-slave.c
@@ -47,9 +47,9 @@ int i2c_slave_register(struct i2c_client *client, i2c_slave_cb_t slave_cb)
 
 	client->slave_cb = slave_cb;
 
-	i2c_lock_bus(client->adapter, I2C_LOCK_ROOT_ADAPTER);
+	i2c_lock_adapter(client->adapter);
 	ret = client->adapter->algo->reg_slave(client);
-	i2c_unlock_bus(client->adapter, I2C_LOCK_ROOT_ADAPTER);
+	i2c_unlock_adapter(client->adapter);
 
 	if (ret) {
 		client->slave_cb = NULL;
@@ -69,9 +69,9 @@ int i2c_slave_unregister(struct i2c_client *client)
 		return -EOPNOTSUPP;
 	}
 
-	i2c_lock_bus(client->adapter, I2C_LOCK_ROOT_ADAPTER);
+	i2c_lock_adapter(client->adapter);
 	ret = client->adapter->algo->unreg_slave(client);
-	i2c_unlock_bus(client->adapter, I2C_LOCK_ROOT_ADAPTER);
+	i2c_unlock_adapter(client->adapter);
 
 	if (ret == 0)
 		client->slave_cb = NULL;
